### ISSUE TITLE
Parsing that depends on context conditions could be postponed

### ISSUE THREAD

**Issue by patrickdalla** [sepinf-inc/IPED#1355]
Parsing that depends on context conditions should be postponed. For example: parsing that use information from other items should expect those items to be processed first, so the processing should be postponed someway.

I thought in implementing an Exception subclass of TikaException, (ContextConditionNotMetException or ContextConditionPending) that could be thrown by the Parser implementation and could be catched by the ParsingTask. If that exception is thrown, the parsing task would put the evidence item in the next priority queue to try to parse again. This way we don't need to change the API.
But still remains a problem: if the item was postponed to reach the last position on the last queue, the parser should choose to parse it anyway, even without the condition met, to parse other informations not dependable of the prior condition. So we have to inform that this another condition is met: "last parsing try".

**Comment by patrickdalla** [comment]
Well it is not so simple as I thought. 
The condition is discovered only in the ParsingTask, so it should be paused and continued latter, in the next priority queue processing.

**Comment by lfcnassif** [comment]
Line 211 is the main one.

**Comment by patrickdalla** [comment]
Yes. I'm there analyzing.

**Comment by patrickdalla** [comment]
Well, i could successfully process my little case. I created a hashmap of "deprioritized items" associated with the respective task that deprioritized the item, so when it is time to reprocess it, it restarts where it stopped. this hashmap was created in class ProcessingQueues (deprioritizedItems).

But the counters became a little confused, show negative times to end.

Tomorrow I will continue.

**Comment by patrickdalla** [comment]
I think that context variables must also be preserved somewhere to be restablished when the item starts to be processed again.

**Comment by lfcnassif** [comment]
What would happen if file A depends on condition C, and file B depends on file A being processed, but condition C never happens?

**Comment by patrickdalla** [comment]
> I think that context variables must also be preserved somewhere to be restablished when the item starts to be processed again.

Thinking better, context variables are of the context, not evidence dependant. Any changes will already be visible to next calls.

**Comment by lfcnassif** [comment]
> 3 - Next try to process "A" will be postponed again, now to the last position of the last queue. This will be the last postpone. 4 - Next try to process "B" will be postponed again, now to the last position of the last queue. This will be the last postpone.

I think B could be postponed before A, if it is saw first. And all items in each processing queue (including the last queue) currently are processed in parallel.

**Comment by patrickdalla** [comment]
"currently are processed in parallel." Well remembered. This can be a problem.

**Comment by lfcnassif** [comment]
I have thought in the past, instead of hard fixing to what queue number each item goes, to automatically build the processing priorities (what item goes to each queue) using annotations in parsers, e.g.: Parser A depends on Parser B results. This would easy to plug parsers as plugins. But as the number of parser dependencies grows and if third party plugin parsers, not under our control, are installed, I think this eventually could cause some cyclic dependency...

So I left the priority configuration as is. One simple improvement would be to externalize it to a configuration file.

**Comment by lfcnassif** [comment]
> [patrickdalla](https://github.com/patrickdalla) closed this as [completed](https://github.com/sepinf-inc/IPED/issues?q=is%3Aissue+is%3Aclosed+archived%3Afalse+reason%3Acompleted) [19 minutes ago](https://github.com/sepinf-inc/IPED/issues/1355#event-7534569663)

Was this intentional?

Anyway, do you think this is needed to implement #281?

**Comment by lfcnassif** [comment]
> As one of the intentions is to extract date and ARES dates are LocalTimes without timezone information, the timezone information should be get from somewhere else.
> 
> Maybe I can use the current fixed scheme.

I think configuring to process Registry files in queue 1 and P2P in queue 2 would be enough.

**Comment by patrickdalla** [comment]
Yes. You are right.

Il giorno gio 6 ott 2022 alle ore 09:40 Luis Filipe Nassif <
***@***.***> ha scritto:

**Comment by patrickdalla** [comment]
If C is met example:

In sequence:
1 - "A" will be postponed to the last queue, as C is not met.
2 - "B" will be postponed as "A" didn't finished.
3 - Next try to process "A" will be postponed again, now to the last
position of the last queue. This will be the last postpone.
5 - Process "D" makes the changes to attend C condition.
4 - Next try to process "B" will be postponed again, now to the last
position of the last queue. This will be the last postpone.
6 - Next try to process "A" will finish successfully.
7 - Next try to process "B" will finish successfully.

Il giorno gio 6 ott 2022 alle ore 08:40 Patrick Bernardina <
***@***.***> ha scritto:

**Comment by lfcnassif** [comment]
> If C is met example: In sequence: 1 - "A" will be postponed to the last queue, as C is not met. 2 - "B" will be postponed as "A" didn't finished. 3 - Next try to process "A" will be postponed again, now to the last position of the last queue. This will be the last postpone. 5 - Process "D" makes the changes to attend C condition. 4 - Next try to process "B" will be postponed again, now to the last position of the last queue. This will be the last postpone. 6 - Next try to process "A" will finish successfully. 7 - Next try to process "B" will finish successfully.

I still think this may not work if B is seen before A and because items in the same queue could be processed concurrently.

**Comment by lfcnassif** [comment]
I suggest leaving this as a future improvement for when it is really needed and when current approach is not enough to implement something.

**Comment by patrickdalla** [comment]
Right. I agree.

Il giorno mar 11 ott 2022 alle ore 09:04 Luis Filipe Nassif <
***@***.***> ha scritto:

**Comment by lfcnassif** [comment]
I'm closing this for now. We can reopen in the future if it is needed.
